tools/shared.py

```
@@ -1749,47 +1754,53 @@ def consider_archive(f):
+      current_archive_group = None
+      for f in files:
+        absolute_path_f = Building.make_paths_absolute(f)
+        if f.startswith('-'):
+          if f in ['--start-group', '-(']:
+            assert current_archive_group is None, 'Nested --start-group, missing --end-group?'
+            current_archive_group = []
+          elif f in ['--end-group', '-)']:
+            assert current_archive_group is not None, '--end-group without --start-group'
+            # rescan the archives in the group until we don't find any more
+            # objects to link.
+            loop_again = True
+            logging.debug('starting archive group loop')
+            while loop_again:
+              loop_again = False
+              for archive in current_archive_group:
+                if consider_archive(archive):
+                  loop_again = True
+            logging.debug('done with archive group loop')
+            current_archive_group = None
           else:
-            # If there are no archives then we can simply link all valid bitcode
-            # files and skip the symbol table stuff.
-            actual_files.append(f)
-      else:
-        # Extract object files from ar archives, and link according to gnu ld semantics
-        # (link in an entire .o from the archive if it supplies symbols still unresolved)
-        consider_archive(absolute_path_f)
-        # If we're inside a --start-group/--end-group section, add to the list
-        # so we can loop back around later.
-        if current_archive_group is not None:
-          current_archive_group.append(absolute_path_f)
-    assert current_archive_group is None, '--start-group without matching --end-group'
+            logging.debug('Ignoring unsupported link flag: %s' % f)
+        elif not Building.is_ar(absolute_path_f):
+          if get_llvm_target() == WASM_OBJ_TARGET:
+            if Building.is_wasm(f):
+              actual_files.append(f)
```

no else? perhaps it should throw an error?